In [1]:
# pip install ucimlrepo

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

# data set source:
from ucimlrepo import fetch_ucirepo, list_available_datasets

In [4]:
heart_disease = fetch_ucirepo(id=45)

# access metadata
# print(heart_disease.metadata.uci_id)
# print(heart_disease.metadata.num_instances)
print(f"Summary:\n{heart_disease.metadata.additional_info.summary}\n")

print(f"Abstract:\n{heart_disease.metadata.abstract}")

Summary:
This database contains 76 attributes, but all published experiments refer to using a subset of 14 of them.  In particular, the Cleveland database is the only one that has been used by ML researchers to date.  The "goal" field refers to the presence of heart disease in the patient.  It is integer valued from 0 (no presence) to 4. Experiments with the Cleveland database have concentrated on simply attempting to distinguish presence (values 1,2,3,4) from absence (value 0).  
   
The names and social security numbers of the patients were recently removed from the database, replaced with dummy values.

One file has been "processed", that one containing the Cleveland database.  All four unprocessed files also exist in this directory.

To see Test Costs (donated by Peter Turney), please see the folder "Costs" 

Abstract:
4 databases: Cleveland, Hungary, Switzerland, and the VA Long Beach


In [5]:
# print(fr"Characteristics: {heart_disease.metadata.characteristics}")
print(fr"Num Rows: {heart_disease.metadata.num_instances}")
print(fr"Num Columns: {heart_disease.metadata.num_features}")
print(fr"Feature Types: {heart_disease.metadata.feature_types}")
print(fr"Target Column(s) Name: {heart_disease.metadata.target_col}")
print(fr"Index Columns Names: {heart_disease.metadata.index_col}")
print(fr"Has Missing Values: {heart_disease.metadata.has_missing_values}")
print(fr"Missing Value Symbol: {heart_disease.metadata.missing_value_symbol}")
print(fr"Data URL: {heart_disease.metadata.data_url}")

Num Rows: 303
Num Columns: 13
Feature Types: ['Categorical', 'Integer', 'Real']
Target Column(s) Name: ['num']
Index Columns Names: None
Has Missing Values: yes
Missing Value Symbol: None
Data URL: https://archive.ics.uci.edu/static/public/45/data.csv


In [6]:
# # print(heart_disease)
# print(type(heart_disease))

# print(type(heart_disease.data.features))

df = heart_disease.data.original
df.columns

Index(['age', 'sex', 'cp', 'trestbps', 'chol', 'fbs', 'restecg', 'thalach',
       'exang', 'oldpeak', 'slope', 'ca', 'thal', 'num'],
      dtype='object')

In [7]:
df.shape

(303, 14)

In [15]:
df_columns = df.columns.tolist() # converst df.columns to a usable list
print(f"All Columns:\n{df_columns} \n")

feature_columns = df_columns.copy()
feature_columns.remove("num")
print(f"Feature Columns:\n{feature_columns}\n")

target_column = df_columns[-1]
print(f"Targe Column:\n{target_column}\n")

All Columns:
['age', 'sex', 'cp', 'trestbps', 'chol', 'fbs', 'restecg', 'thalach', 'exang', 'oldpeak', 'slope', 'ca', 'thal', 'num'] 

Feature Columns:
['age', 'sex', 'cp', 'trestbps', 'chol', 'fbs', 'restecg', 'thalach', 'exang', 'oldpeak', 'slope', 'ca', 'thal']

Targe Column:
num



In [19]:
df.dtypes

age           int64
sex           int64
cp            int64
trestbps      int64
chol          int64
fbs           int64
restecg       int64
thalach       int64
exang         int64
oldpeak     float64
slope         int64
ca          float64
thal        float64
num           int64
dtype: object

In [29]:
df.head()

,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,num
0,63,1,1,145,233,1,2,150,0,2.3,3,0.0,6.0,0
1,67,1,4,160,286,0,2,108,1,1.5,2,3.0,3.0,2
2,67,1,4,120,229,0,2,129,1,2.6,2,2.0,7.0,1
3,37,1,3,130,250,0,0,187,0,3.5,3,0.0,3.0,0
4,41,0,2,130,204,0,2,172,0,1.4,1,0.0,3.0,0


In [37]:
feature_explain_df = pd.read_csv("Features_List.csv")
feature_explain_df.head()

,Name,Role,Type,Units,Description,Missing Values
0,age,feature,integer,years,NaN,No
1,sex,feature,categorical,NaN,NaN,No
2,cp,feature,categorical,NaN,NaN,No
3,trestbps,feature,integer,mm Hg,resting blood pressure (on admission to the ho...,No
4,chol,feature,integer,mg/dl,serum cholestoral,No


The creators of the dataset have already converted the categorical features into numeric values. I will need to find what each numeric value represnts.

In [42]:
# get the ranges of each column
for col in df_columns:
    print(f"{col}: {df[col].min()} - {df[col].max()}")

age: 29 - 77
sex: 0 - 1
cp: 1 - 4
trestbps: 94 - 200
chol: 126 - 564
fbs: 0 - 1
restecg: 0 - 2
thalach: 71 - 202
exang: 0 - 1
oldpeak: 0.0 - 6.2
slope: 1 - 3
ca: 0.0 - 3.0
thal: 3.0 - 7.0
num: 0 - 4


## Attribute documentation

- age: age in years
- sex: sex (1 = male; 0 = female)
- painloc: chest pain location (1 = substernal; 0 = otherwise)
- cp: chest pain type
  - Value 1: typical angina
  - Value 2: atypical angina
  - Value 3: non-anginal pain
  - Value 4: asymptomatic
- trestbps: resting blood pressure (in mm Hg on admission to the hospital)
- htn
- chol: serum cholestoral in mg/dl
- fbs: (fasting blood sugar > 120 mg/dl) (1 = true; 0 = false)
- thalach: maximum heart rate achieved
- exang: exercise induced angina (1 = yes; 0 = no)
- oldpeak = ST depression induced by exercise relative to rest
- slope: the slope of the peak exercise ST segment
  - Value 1: upsloping
  - Value 2: flat
  - Value 3: downsloping
- ca: number of major vessels (0-3) colored by flourosopy
- thal: 3 = normal; 6 = fixed defect; 7 = reversable defect
- num: diagnosis of heart disease (angiographic disease status)
  - Value 0: < 50% diameter narrowing
  - Value 1: > 50% diameter narrowing
  - (in any major vessel: attributes 59 through 68 are vessels)|